pkg/identity/codefresh/principal.go

```diff
@@ -0,0 +1,166 @@
+	platformURL string
+}
+
+func (w workflowPrincipal) Name(_ context.Context) string {
+	return w.subject
+}
+
+func WorkflowPrincipalFromIDToken(_ context.Context, token *oidc.IDToken) (identity.Principal, error) {
+	var claims struct {
+		AccountID         string `json:"account_id"`
+		AccountName       string `json:"account_name"`
+		PipelineID        string `json:"pipeline_id"`
+		PipelineName      string `json:"pipeline_name"`
+		WorkflowID        string `json:"workflow_id"`
+		Initiator         string `json:"initiator"`
+		SCMRepoURL        string `json:"scm_repo_url"`
+		SCMUsername       string `json:"scm_user_name"`
+		SCMRef            string `json:"scm_ref"`
+		SCMPullRequestRef string `json:"scm_pull_request_target_branch"`
+		RunnerEnvironment string `json:"runner_environment"`
+		PlatformURL       string `json:"platform_url"`
+	}
+
+	if err := token.Claims(&claims); err != nil {
+		return nil, err
+	}
+
+	if claims.AccountID == "" {
+		return nil, errors.New("missing account_id in token")
+	}
+
+	if claims.PipelineID == "" {
+		return nil, errors.New("missing pipeline_id in token")
+	}
+
+	if claims.WorkflowID == "" {
+		return nil, errors.New("missing workflow_id in token")
+	}
+
+	// Set default platform url in case it is missing in the token
+	if claims.PlatformURL == "" {
+		claims.PlatformURL = DefaultPlatformURL
+	}
+
+	return &workflowPrincipal{
+		subject:                    token.Subject,
+		issuer:                     token.Issuer,
+		accountID:                  claims.AccountID,
+		accountName:                claims.AccountName,
+		pipelineID:                 claims.PipelineID,
+		pipelineName:               claims.PipelineName,
+		workflowID:                 claims.WorkflowID,
+		initiator:                  claims.Initiator,
+		scmUsername:                claims.SCMUsername,
+		scmRepoURL:                 claims.SCMRepoURL,
+		scmRef:                     claims.SCMRef,
+		scmPullRequestTargetBranch: claims.SCMPullRequestRef,
+		runnerEnvironment:          claims.RunnerEnvironment,
+		platformURL:                claims.PlatformURL,
+	}, nil
+}
+
+func (w workflowPrincipal) Embed(_ context.Context, cert *x509.Certificate) error {
+
+	baseURL, err := url.Parse(w.platformURL)
+
+	if err != nil {
+		return err
+	}
+
+	// Set SAN to the <platform url>/<account name>/<pipeline name>:<account id>/<pipeline id> - for example https://g.codefresh.io/codefresh-account/oidc-test/get-token:628a80b693a15c0f9c13ab75/65e5a53e52853dc51a5b0cc1
+	// In Codefresh account names and pipeline names may be changed where as IDs do not.
+	// This pattern will give users the possibility to verify the signature using various forms of `cosign verify --certificate-identity-regexp` i.e https://g.codefresh.io/codefresh-account/oidc-test/get-token:* or https://g.codefresh.io/*:628a80b693a15c0f9c13ab75/65e5a53e52853dc51a5b0cc1
+	cert.URIs = []*url.URL{baseURL.JoinPath(w.accountName, fmt.Sprintf("%s:%s/%s", w.pipelineName, w.accountID, w.pipelineID))}
```

Comment:
If account name and pipeline are mutable, should this be composed on only the account and pipeline IDs?

Are pipeline IDs unique across all pipelines, or just unique within an account? If the former, then this could be composed of just a pipeline ID, or maybe use account name + pipeline ID so it's more human readable?

Reply:
Pipeline IDs are unique across accounts. Problem is that it's not very intuitively accessible to users from Codefresh UI, and even more so it's not accessible to unauthenticated users in case of public builds. But I think the usage of `--certificate-identity-regexp` allows complete flexibility in the proposed way of structuring the SAN. It allows the verifier to decide which parameters they trust.

Reply:
Account name is indeed there - It's the first member in `JoinPath`

Reply:
whoops, sorry, i was looking at the latter half!